cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/RemoteValidationTask.java

```diff
@@ -0,0 +1,114 @@
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.macro.MacroEvaluator;
+import io.cdap.cdap.api.macro.MacroParserOptions;
+import io.cdap.cdap.api.service.worker.RunnableTask;
+import io.cdap.cdap.api.service.worker.RunnableTaskContext;
+import io.cdap.cdap.api.service.worker.TaskSystemAppContext;
+import io.cdap.cdap.etl.common.BasicArguments;
+import io.cdap.cdap.etl.common.ConnectionMacroEvaluator;
+import io.cdap.cdap.etl.common.DefaultMacroEvaluator;
+import io.cdap.cdap.etl.common.OAuthMacroEvaluator;
+import io.cdap.cdap.etl.common.SecureStoreMacroEvaluator;
+import io.cdap.cdap.etl.proto.v2.validation.StageValidationRequest;
+import io.cdap.cdap.etl.proto.v2.validation.StageValidationResponse;
+import io.cdap.cdap.internal.io.SchemaTypeAdapter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Task for executing validation requests remotely
+ */
+public class RemoteValidationTask implements RunnableTask {
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+    .serializeNulls()
+    .create();
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteValidationTask.class);
+
+  @Override
+  public void run(RunnableTaskContext context) throws Exception {
+    TaskSystemAppContext systemAppContext = context.getRunnableTaskSystemAppContext();
+    RemoteValidationRequest remoteValidationRequest = GSON.fromJson(context.getParam(), RemoteValidationRequest.class);
+    String namespace = remoteValidationRequest.getNamespace();
+    String serializedRequest = remoteValidationRequest.getSerializedRequest();
+    StageValidationRequest validationRequest;
+    try {
+      validationRequest = GSON.fromJson(serializedRequest,
+                                        StageValidationRequest.class);
+      validationRequest.validate();
+    } catch (JsonSyntaxException e) {
+      throw new IllegalArgumentException(String.format("Unable to decode request body %s", serializedRequest), e);
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException("Invalid stage config", e);
+    }
+
+
+    Map<String, String> arguments = Collections.emptyMap();
+
+    // Fetch preferences for this instance and namespace and use them as program arguments if the user selects
+    // this option.
+    if (validationRequest.getResolveMacrosFromPreferences()) {
+      try {
+        arguments = systemAppContext.getPreferencesForNamespace(namespace, true);
+      } catch (IllegalArgumentException iae) {
+        // If this method returns IllegalArgumentException, it means the namespace doesn't exist.
+        // If this is the case, we return a 404 error.
+        throw new IllegalArgumentException(String.format("Namespace '%s' does not exist", namespace), iae);
+      }
+    }
+
+    Map<String, String> finalArguments = arguments;
```

Comment:
why is this assignment needed, could we use `arguments` map?

Reply:
arguments is not final, and the lambda requires final/effectively final args(compilation error).

Reply:
N/A with other changes.